xarray/core/groupby.py

```diff
@@ -467,15 +519,21 @@ def factorize(self) -> EncodedGroups:
         # Restore these after the raveling
         broadcasted_masks = broadcast(*masks)
         mask = functools.reduce(np.logical_or, broadcasted_masks)  # type: ignore[arg-type]
-        _flatcodes[mask] = -1
+        _flatcodes = where(mask, -1, _flatcodes)
 
         full_index = pd.MultiIndex.from_product(
             (grouper.full_index.values for grouper in groupers),
             names=tuple(grouper.name for grouper in groupers),
         )
-        # Constructing an index from the product is wrong when there are missing groups
-        # (e.g. binning, resampling). Account for that now.
-        midx = full_index[np.sort(pd.unique(_flatcodes[~mask]))]
+        # This will be unused when grouping by dask arrays, so skip..
+        if not is_chunked_array(_flatcodes):
+            # Constructing an index from the product is wrong when there are missing groups
+            # (e.g. binning, resampling). Account for that now.
+            midx = full_index[np.sort(pd.unique(_flatcodes[~mask]))]
```

Comment:
Why not np.unique? You'll get the results sorted then.

Reply:
`np.unique` sorts first. This can be quite slow if `_flatcodes` is large, which it can be,

Reply:
That's interesting, thank you!